sharktank/tests/models/llama/attention_test.py

```diff
@@ -0,0 +1,151 @@
+# Copyright 2024 Advanced Micro Devices, Inc
+#
+# Licensed under the Apache License v2.0 with LLVM Exceptions.
+# See https://llvm.org/LICENSE.txt for license information.
+# SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+import unittest
+from typing import List
+
+import torch
+
+from sharktank.models.llama.testing import *
+from sharktank.layers.rotary_embedding import RotaryEmbeddingLayer
+from sharktank.models.llama.llama import PagedLlamaAttentionBlock, PagedKVCache
+from sharktank import ops
+
+from transformers.models.llama.modeling_llama import (
+    LlamaAttention,
+    LlamaMLP,
+    LlamaRMSNorm,
+    LlamaDecoderLayer,
+)
+from transformers.models.llama.configuration_llama import LlamaConfig
+
+
+class AttentionBlockTest(unittest.TestCase):
+    def test(self):
+        torch.set_default_dtype(torch.float32)
+        block_index = 0
+        seq_len = 13
+        head_count = 32
+        head_dim = 100
+        hidden_size = 3200
+        ffn_dim = 8640
+        head_count_kv = 32
+        block_seq_stride = 1
+        rms_epsilon = 0.01
+        rope_dimension_count = 100
+        max_seq_len = 2048
+        attention_block_theta = make_attention_block_theta(
+            feature_dim=head_count * head_dim, ffn_dim=ffn_dim, dtype=torch.float32
+        )
+        paged_kv_cache = PagedKVCache(
+            transformer_block_count=head_count,
+            attn_head_count=head_count,
+            attn_head_dim=head_dim,
+            cache_partition_count=2,  # One for each of K/V.
+            block_seq_stride=block_seq_stride,
+            device="cpu",
+            dtype=torch.float32,
+        )
+        attention_block = PagedLlamaAttentionBlock(
+            theta=attention_block_theta,
+            block_index=block_index,
+            cache=paged_kv_cache,
+            head_count=head_count,
+            head_dim=head_dim,
+            head_count_kv=head_count_kv,
+            rms_epsilon=rms_epsilon,
+            hf=True,
```

Comment:
So this isn't a direct call to huggingface, it's simulating the same math that huggingface uses?

Reply:
It's using all of HF llama layers but we are integrating/building ourselves I believe

Reply:
actually not sure anymore what you were pointing at haha, but yes this PR should introduce the math that HF uses for rope embedding

Reply:
Yep, this attention block is our own sharktank implementation. We are comparing its outputs to its corresponding equivalent in huggingface (decoder layer)